### PR TITLE
State Recovery Swapd Implementation

### PR DESCRIPTION
**Pull request by TheCharlatan**
This adds swapd state serialization, checkpointing and recovery.

The current workflow for restoration is only for cases where farcasterd is restarted. Peer reconnection is also not handled.

I left out of scope for now checking if we are overwriting a later state on recovery. For the current workflow it does not add much value and I don't want this pull request to have an even larger scope.

### CODE REVIEW

**Comment by TheCharlatan** [comment]
I had a successful state restoration with the latest commit. I made and took an offer, funded it, terminated farcasterd, started farcasterd, restored the checkpoint and eventually got the amount refunded again. I'm going to keep on polishing and testing and on this pull request while I wait for review and merge of the others.

File: src/farcasterd/runtime.rs [hunk 1]
```diff
@@ -889,6 +909,93 @@ impl Runtime {
...
+                    &mut self.spawning_services,
+                    &self.syncer_services,
+                    &mut self.syncer_clients,
+                    Coin::Monero,
+                    public_offer.offer.network,
+                    swap_id,
+                    &self.config,
+                )?;
+
+                let _child = launch(
```
Comment (TheCharlatan) [inline src/farcasterd/runtime.rs:982]: Check if swapd is already running before launching a new swap.

File: src/swapd/runtime.rs [hunk 2]
```diff
@@ -1919,6 +2172,110 @@ impl Runtime {
...
+                    state,
+                    last_msg,
+                    enquirer,
+                    temporal_safety,
+                    txs,
+                    txids,
+                    pending_requests,
+                    xmr_addr_addendum,
+                }) => {
+                    info!("{} | Restoring swap", swap_id);
```
Comment (TheCharlatan) [inline src/swapd/runtime.rs:2195]: Check if we are restoring to a later state, and are not overwriting an older.

File: doc/staterecovery_sequencediagram.txt [hunk 3]
```diff
@@ -1,16 +1,21 @@
...
 participant farcasterd
 participant cli
 
 == State Recovery Procedure
 entryspacing 0.8
+farcasterd -> walletd : launch
 farcasterd -> databased : launch
 cli -> databased : RetrieveAllCheckpointInfo
-databased -> cli : CheckpointList
+databased -> farcasterd : CheckpointList
```
Comment (zkao) [inline doc/staterecovery_sequencediagram.txt:15]: this is inconsistent with [main](https://github.com/farcaster-project/farcaster-node/blob/main/doc/staterecovery_sequencediagram.svg)

File: src/walletd/runtime.rs [hunk 4]
```diff
@@ -1510,6 +1603,8 @@ impl Runtime {
                         source,
                         Request::SweepXmrAddress(sweep_keys),
                     )?;
+                } else {
+                    error!("\n\nnot an Alice wallet!\n\n");
```
Comment (TheCharlatan) [inline src/walletd/runtime.rs:1607]: Silly log

File: src/rpc/request.rs [hunk 5]
```diff
@@ -740,10 +744,11 @@ pub struct Checkpoint {
 }
 
 #[derive(Clone, Debug, Display, StrictDecode, StrictEncode)]
-#[display(Debug)]
 pub enum CheckpointState {
-    CheckpointWalletAlice(AliceState),
-    CheckpointWalletBob(BobState),
+    #[display("Checkpoint Wallet")]
+    CheckpointWallet(CheckpointWallet),
+    #[display("Checkpoint Wallet")]
```
Comment (zkao) [inline src/rpc/request.rs:750]: s/Wallet/Swap

**Review by zkao — COMMENTED** [review]
Since there are proper tests written already testing the added functionality, I would like to explore and test more on a fork of this branch

I guess I understood enough of the overall architecture, and the most of the intends on the code

File: src/walletd/runtime.rs [hunk 6]
```diff
@@ -1555,21 +1650,15 @@ impl Runtime {
...
 
             Request::Checkpoint(request::Checkpoint { swap_id, state }) => match state {
-                CheckpointState::CheckpointWalletAlice(alice_state) => {
-                    info!("Restoring alice wallet for swap {}", swap_id);
+                CheckpointState::CheckpointWallet(CheckpointWallet { wallet, xmr_addr }) => {
+                    info!("Restoring wallet for swap {}", swap_id);
                     if !self.wallets.contains_key(&swap_id) {
-                        self.wallets.insert(swap_id, Wallet::Alice(alice_state));
-                    } else {
-                        error!("Did not restore, the runtime already contains an Alice wallet for swap {}", swap_id)
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/walletd/runtime.rs:1563]: There is no distinguishing between restoring Alice or Bob wallets anymore.

File: src/swapd/runtime.rs [hunk 7]
```diff
@@ -1919,6 +2172,110 @@ impl Runtime {
...
+                        let task = self
+                            .syncer_state
+                            .watch_tx_btc(txid.clone(), tx_label.clone());
+                        endpoints.send_to(
+                            ServiceBus::Ctl,
+                            self.identity(),
+                            self.syncer_state.bitcoin_syncer(),
+                            Request::SyncerTask(task),
+                        )?;
+                    }
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2245]: Yes, it's defensive to ensure that all known transactions are re-watched. I agree this should be unified. The two HashMaps can be zipped on their tx label and txid.

File: src/farcasterd/runtime.rs [hunk 8]
```diff
@@ -916,17 +942,89 @@ impl Runtime {
...
-                endpoints.send_to(
-                    ServiceBus::Ctl,
+
+                // check if swapd is not running
+                if endpoints
+                    .send_to(
+                        ServiceBus::Msg,
+                        ServiceId::Farcasterd,
+                        ServiceId::Swap(swap_id.clone()),
+                        Request::Hello,
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/farcasterd/runtime.rs:952]: I guess we can add a `ConnectionTest`?

**Comment by TheCharlatan** [comment]
Thank you for the great review zkao. I will attempt to address your comments first in writing and will follow-up with some code later.

File: src/farcasterd/runtime.rs [hunk 9]
```diff
@@ -916,17 +942,89 @@ impl Runtime {
...
+                            info: "Cannot restore a checkpoint into a running swap.".to_string(),
+                        }),
+                    )?;
+                    return Ok(());
+                }
+
+                let CheckpointEntry {
+                    public_offer,
+                    trade_role,
+                    ..
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/farcasterd/runtime.rs:971]: It's nice to have these, so you can retrieve them when you do `GetInfo` on the swap.

File: src/swapd/runtime.rs [hunk 10]
```diff
@@ -203,11 +211,168 @@ pub struct Runtime {
...
+            .unmarshall(Cursor::new(Vec::<u8>::strict_decode(&mut d)?))
+            .unwrap())
+            .clone();
+        Ok(PendingRequest {
+            dest,
+            bus_id,
+            request,
+        })
+    }
+}
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:251]: AFAICT `PendingRequest` is only used in swapd, so I think it's fine to keep it there. The contents of the Request file should be variants of the `Request` enum, which `PendingRequest` is not.

File: src/swapd/runtime.rs [hunk 11]
```diff
@@ -203,11 +211,168 @@ pub struct Runtime {
...
+            last_msg,
+            enquirer,
+            xmr_addr_addendum,
+            temporal_safety,
+            txs,
+            txids,
+            pending_requests,
+        })
+    }
+}
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:374]: It should once it only contains data that is contained in `swap_state`.

File: src/swapd/runtime.rs [hunk 12]
```diff
@@ -1178,10 +1365,10 @@ impl Runtime {
                             let swap_success_req = Request::SwapOutcome(success);
                             self.send_ctl(endpoints, ServiceId::Wallet, swap_success_req.clone())?;
                             self.send_ctl(endpoints, ServiceId::Farcasterd, swap_success_req)?;
-                            // remove txs to invalidate outdated states
+                            // remove txs from outdated states
+                            self.txs.remove(&TxLabel::Lock);
                             self.txs.remove(&TxLabel::Cancel);
                             self.txs.remove(&TxLabel::Refund);
-                            self.txs.remove(&TxLabel::Buy);
                             self.txs.remove(&TxLabel::Punish);
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:1372]: Yeah, this was a bad idea. I now understand what you were doing here. I'll revert this change.

File: src/swapd/runtime.rs [hunk 13]
```diff
@@ -1479,9 +1665,8 @@ impl Runtime {
                                     && self.txs.contains_key(&TxLabel::Refund) =>
                             {
                                 trace!("here Bob publishes refund tx");
-                                let (tx_label, refund_tx) =
-                                    self.txs.remove_entry(&TxLabel::Refund).unwrap();
-                                self.broadcast(refund_tx, tx_label, endpoints)?;
+                                let refund_tx = self.txs.get(&TxLabel::Refund).unwrap().clone();
+                                self.broadcast(refund_tx, TxLabel::Refund, endpoints)?;
                             }
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:1670]: For the Refund and Cancel tx I think it is safe enough to `remove_entry`, they are in the checkpoint pre buy anyway.

For the other txs, let me go over these changes again. I feared the execution flow might be unable to progress if we don't ensure that these txs are not removed before checkpointing.

File: src/swapd/runtime.rs [hunk 14]
```diff
@@ -1527,14 +1712,15 @@ impl Runtime {
...
+                                // remove txs from outdated states
+                                self.txs.remove(&TxLabel::Lock);
                                 self.txs.remove(&TxLabel::Buy);
                                 self.txs.remove(&TxLabel::Cancel);
                                 self.txs.remove(&TxLabel::Punish);
                             }
                             TxLabel::Buy
                                 if self.temporal_safety.final_tx(*confirmations, Coin::Bitcoin)
-                                    && self.state.a_refundsig()
-                                    && self.state.a_buy_published() =>
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:1537]: > I think I get it. Checkpoint is not updated after checkpoint, but please confirm!

Yeah, that is it. 

I need to understand what we are protecting against by checking `a_buy_published`? How can this case be hit in the first place without it being published?

File: src/swapd/runtime.rs [hunk 15]
```diff
@@ -1919,6 +2172,110 @@ impl Runtime {
...
+                    trace!("Watching transactions");
+                    for (tx_label, tx) in txs.iter() {
+                        let task = self.syncer_state.watch_tx_btc(tx.txid(), tx_label.clone());
+                        endpoints.send_to(
+                            ServiceBus::Ctl,
+                            self.identity(),
+                            self.syncer_state.bitcoin_syncer(),
+                            Request::SyncerTask(task),
+                        )?;
+                    }
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2234]: I will comb over the need to store the txs again. Maybe we can indeed get rid of their inclusion in the checkpoint.

File: src/walletd/runtime.rs [hunk 16]
```diff
@@ -112,6 +113,29 @@ use strict_encoding::{
     strategies::HashFixedBytes, strict_encode_list, Strategy, StrictDecode, StrictEncode,
 };
 
+#[derive(Clone, Debug)]
+pub struct CheckpointWallet {
+    pub wallet: Wallet,
+    pub xmr_addr: monero::Address,
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/walletd/runtime.rs:119]: Same with the btc_addr? Can this go in a future PR?

File: src/swapd/runtime.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs]: Reverted in 307d529a4accc8f840aa535941c1c91b32b02956

File: src/farcasterd/runtime.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (zkao) [reply src/farcasterd/runtime.rs:952]: sure

File: src/farcasterd/runtime.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (zkao) [reply src/farcasterd/runtime.rs:971]: my point is not about not having them. It's about not passing them as command line args, passing them through the Ctl bus instead

File: src/walletd/runtime.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (zkao) [reply src/walletd/runtime.rs:119]: sure, just wondering

File: src/walletd/runtime.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (zkao) [reply src/walletd/runtime.rs:1563]: i meant, why remove logging completely? that is an error if it happens

File: src/swapd/runtime.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs]: I combed over these changes again and was able to remove all of them. Thank you for catching this! 86606176f66abbd145085b1507585a11628937a8

File: src/swapd/runtime.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2234]: I removed watching the tx's altogether.

File: src/swapd/runtime.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs]: I remove this in 70dfa6c77386c44048a738bbc6ac52a591a42c32 .

File: src/walletd/runtime.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/walletd/runtime.rs:1563]: Added in b00f632c2a98b0218cddbf35ff9a788efc34997f

File: src/swapd/runtime.rs [hunk 17]
```diff
@@ -203,11 +211,168 @@ pub struct Runtime {
...
+                })?;
+                Ok(acc)
+            });
+        len = match res {
+            Ok(val) => Ok(val),
+            Err(err) => Err(strict_encoding::Error::DataIntegrityError(format!(
+                "{}",
+                err
+            ))),
+        }?;
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:292]: I did not find a derive for HashMap, so I rolled this myself. I guess we can create a wrapper type for a HashMap, implement the strict encoding for it and then use it here.

**Pull request merged**